File: Data Structures & Algorithms/matrixBFS/submission-2.java
public class Solution {

    public int shortestPath(int[][] grid) {
        Queue<int[]> queue = new LinkedList<>();
        if (grid[0][0] == 0) {
            queue.add(new int[]{0, 0, 0});
        }
        int res = Integer.MAX_VALUE;
        Set<String> visited = new HashSet<>();

        while (!queue.isEmpty()) {
            int[] poll = queue.poll();
            int row = poll[0];
            int col = poll[1];
            int pathCount = poll[2];

            if (row == grid.length - 1 && col == grid[0].length - 1) {
                res = Math.min(res, pathCount);
                continue;
            }

            if (visited.contains(row + ":" + col) || !isInRange(grid, row, col) || grid[row][col] == 1) {
                continue;
            }

            visited.add(row + ":" + col);
            queue.add(new int[]{row, col + 1, pathCount + 1});
            queue.add(new int[]{row + 1, col, pathCount + 1});
            queue.add(new int[]{row, col - 1, pathCount + 1});
            queue.add(new int[]{row - 1, col, pathCount + 1});
        }

        return res == Integer.MAX_VALUE ? -1 : res;
    }

    private boolean isInRange(int[][] grid, int row, int col) {
        return row >= 0 && col >= 0 && row < grid.length && col < grid[0].length;
    }
}

File: Data Structures & Algorithms/rotting-fruit/submission-2.kt
class Solution {
    fun orangesRotting(grid: Array<IntArray>): Int {
        val queue = LinkedList<Triple<Int, Int, Int>>()
        val visited = HashSet<String>()
        var res = 0

        for (row in 0..<grid.size) {
            for (col in 0..<grid[0].size) {
                if (grid[row][col] == 2) queue.add(Triple(row, col, 0))
            }
        }

        while (queue.isNotEmpty()) {
            var pop = queue.pop()
            var row = pop.component1()
            var col = pop.component2()
            var minutes = pop.component3()

            if (row < 0 || row >= grid.size || col < 0 || col >= grid[0].size || visited.contains("$row - $col")
                || grid[row][col] == 0
            ) continue
            visited.add("$row - $col")
            grid[row][col] = 2
            res = max(res, minutes)
            queue.add(Triple(row, col + 1, minutes + 1))
            queue.add(Triple(row + 1, col, minutes + 1))
            queue.add(Triple(row, col - 1, minutes + 1))
            queue.add(Triple(row - 1, col, minutes + 1))
        }

        for (nums in grid) {
            for (n in nums) {
                if (n == 1) return -1
            }
        }

        return res
    }
}